### PR TITLE
Metre analysis w/ sinaeresis and synalephas

### PR DESCRIPTION
**Pull request by alvp**

### CODE REVIEW

File: src/rantanplan/core.py [hunk 1]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+    sinaeresis_words = word[:]  # create a copy since we're going to pop()
+    prosodic_syllables = []
+    for idx, key in enumerate(sinaeresis_words):
+        if not key.get('has_sinaeresis'):
+            prosodic_syllables.append(sinaeresis_words[idx])
+        else:
+            next_syl = sinaeresis_words.pop(idx + 1)
+            prosodic_syllables.append({'is_stressed': (key['is_stressed'] or next_syl['is_stressed']),
+                                       'syllable': "".join([key['syllable'], next_syl['syllable']]),
+                                       'has_sinaeresis': True})
```
Comment (versae) [inline src/rantanplan/core.py]: Rewritten as
```python
            prosodic_syllables.append({
                'is_stressed': (key['is_stressed'] or next_syl['is_stressed']),
                'syllable': "".join([key['syllable'], next_syl['syllable']]),
                'has_sinaeresis': True
            })
```

File: src/rantanplan/core.py [hunk 2]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+                prosodic_line.append({'is_stressed': (syllable.get('is_stressed') or next_syllable.get('is_stressed')),
+                                      'syllable': "".join([syllable.get('syllable'), next_syllable.get('syllable')]),
+                                      'has_synalepha': True})
+    return prosodic_line
+
+
+def get_sinaeresis(word):
+    sinaeresis_words = word[:]  # create a copy since we're going to pop()
+    prosodic_syllables = []
+    for idx, key in enumerate(sinaeresis_words):
```
Comment (versae) [inline src/rantanplan/core.py]: Same thing, it feels like a proper for loop could avoid the use of the copy and the `pop`

File: src/rantanplan/core.py [hunk 3]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+            prosodic_syllables.append(sinaeresis_words[idx])
+        else:
+            next_syl = sinaeresis_words.pop(idx + 1)
+            prosodic_syllables.append({'is_stressed': (key['is_stressed'] or next_syl['is_stressed']),
+                                       'syllable': "".join([key['syllable'], next_syl['syllable']]),
+                                       'has_sinaeresis': True})
+    return prosodic_syllables
+
+
+def get_metre(poem):
```
Comment (versae) [inline src/rantanplan/core.py]: Missing documentation.

File: src/rantanplan/core.py [hunk 4]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+    for lines in poem:
```
Comment (versae) [inline src/rantanplan/core.py]: `for line in poem`? Should `poem` be renamed as `lines`?

File: src/rantanplan/core.py [hunk 5]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+        metre = ""
```
Comment (versae) [inline src/rantanplan/core.py]: Prefer `metre = []`, `metre.append(...)`, and `"".join(metre)` when possible.

File: src/rantanplan/core.py [hunk 6]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+        token_list = lines['tokens']
+        sinaeresis_tokens = {'sinaeresis_tokens': []}
+        prosodic_tokens = {'prosodic_tokens': []}
```
Comment (versae) [inline src/rantanplan/core.py]: Not sure I understand these two lines. Isn't it just the same than having
```python
        sinaeresis_tokens = []
        prosodic_tokens = []
```

File: src/rantanplan/core.py [hunk 7]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+        for tokens in token_list:
+            if tokens.get('word'):
+                sinaeresis_tokens['sinaeresis_tokens'].append(get_sinaeresis(tokens.get('word')))
+        synalepha_token_list = get_synalephas(sinaeresis_tokens['sinaeresis_tokens'])
+        for tokens in synalepha_token_list:
```
Comment (versae) [inline src/rantanplan/core.py]: `for token in synalepha_token_list`?

File: src/rantanplan/core.py [hunk 8]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+            if tokens.get('syllable'):
+                prosodic_tokens['prosodic_tokens'].append(tokens)
+        lines.update(prosodic_tokens)
+        for syllable in synalepha_token_list:
+            if syllable['is_stressed']:
+                metre += '+'
+            else:
+                metre += '-'
+        last_word = sinaeresis_tokens['sinaeresis_tokens'][-1]
+        print('last word: ', last_word)
```
Comment (versae) [inline src/rantanplan/core.py]: Remove.

File: src/rantanplan/core.py [hunk 9]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+        last_syllable_stress = last_word[-1].get('is_stressed')
+        if last_syllable_stress:
+            metre += '-'
+        elif len(last_word) >= 3:  # Esdrújula
```
Comment (versae) [inline src/rantanplan/core.py]: Is there a term in English for `Esdrújula`? If so, add it as well, please.

File: src/rantanplan/core.py [hunk 10]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+            third_from_last_syllable_stress = last_word[-3].get('is_stressed')
+            if third_from_last_syllable_stress:
+                metre = metre[:-1]
+        elif len(last_word) > 3:  # Sobreesdrújula
```
Comment (versae) [inline src/rantanplan/core.py]: Same for `Sobreesdrújula`

File: src/rantanplan/core.py [hunk 11]
```diff
@@ -310,6 +376,7 @@ def get_scansion(text, rhyme_analysis=False):
             seen_tokens.append(token)
     if len(seen_tokens) > 0:
         lines.append({"tokens": get_syllables(seen_tokens)})
+    lines = get_metre(lines)
```
Comment (versae) [inline src/rantanplan/core.py]: I think this code can be much clearer if we reuse the existing structure. Something like:
```python
    if len(seen_tokens) > 0:
        metrical_pattern, prosodic_tokens = get_metre(seen_tokens)
        lines.append({
            "tokens": get_syllables(seen_tokens),
            "prosodic_tokens": prosodic_tokens,
            "metre": metrical_pattern,
        })
```
That would require to change the function `get_metre` slightly.

File: src/rantanplan/core.py [hunk 12]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
             and second_syllable['syllable'][0] in LIAISON_SECOND_PART)
 
 
+def get_synalephas(words):
+    prosodic_line = []
+    word_list = copy.deepcopy(words)
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/core.py]: I can review the code and change it, I almost went crazy writing that and the get_sinaeresis function. About deepcopy, for the assignment of that specific variable indeed is needed, copy still modifies the original list.

File: src/rantanplan/core.py [hunk 13]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+    for idx, syllables_list in enumerate(word_list):
+        for syllable in syllables_list:
+            if not syllable.get('has_synalepha'):
+                prosodic_line.append(syllable)
+            else:
+                next_syllable = word_list[idx + 1].pop(0)
+                prosodic_line.append({'is_stressed': (syllable.get('is_stressed') or next_syllable.get('is_stressed')),
+                                      'syllable': "".join([syllable.get('syllable'), next_syllable.get('syllable')]),
+                                      'has_synalepha': True})
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/core.py]: Yep, changed it

File: src/rantanplan/core.py [hunk 14]
```diff
@@ -98,6 +99,71 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+    return prosodic_line
+
+
+def get_sinaeresis(word):
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/core.py]: Alredady done

File: src/rantanplan/core.py (hunk 1 above)
Comment (alvp) [reply src/rantanplan/core.py]: Done

File: src/rantanplan/core.py (hunk 8 above)
Comment (alvp) [reply src/rantanplan/core.py]: :upside_down_face:

File: src/rantanplan/core.py (hunk 9 above)
Comment (alvp) [reply src/rantanplan/core.py]: Yes, good old proparoxytone

File: src/rantanplan/core.py (hunk 10 above)
Comment (alvp) [reply src/rantanplan/core.py]: preantepenultimate

File: src/rantanplan/core.py (hunk 11 above)
Comment (alvp) [reply src/rantanplan/core.py]: Can we add it as an issue for future work on this?

File: src/rantanplan/core.py (hunk 7 above)
Comment (alvp) [reply src/rantanplan/core.py]: changed to singular

File: src/rantanplan/core.py [hunk 15]
```diff
@@ -44,12 +44,12 @@
...
     #10: Explicit hiatus with umlaut vowels, second part
     ([äëïöü][a-z])|
-    #12: any char
+    #11: any char
     ([a-záéíóúñ])
 """, re.I | re.U | re.VERBOSE)  # re.VERBOSE is needed to be able to catch the regex group
 
 """
-Metrical Analysis
+rhythmical Analysis
```
Comment (versae) [inline src/rantanplan/core.py]: Why the change? :/ Also, please, capitalize.

File: src/rantanplan/core.py [hunk 16]
```diff
@@ -99,69 +99,96 @@ def have_prosodic_liaison(first_syllable, second_syllable):
             and second_syllable['syllable'][0] in LIAISON_SECOND_PART)
 
 
-def get_synalephas(words):
+def get_synalephas(sinaereris_words_dictionary):
+    """
+    Gets a list of dictionaries for each word on each line of the poem
+    and joins the syllables to create phonologic syllables according to its
```
Comment (versae) [inline src/rantanplan/core.py]: I've seen you've changed `phonologic` to `phonological` in some places. Let's try to keep some consistency with naming.

File: src/rantanplan/core.py [hunk 17]
```diff
@@ -99,69 +99,96 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+    """
     prosodic_line = []
-    word_list = copy.deepcopy(words)
+    word_list = copy.deepcopy(sinaereris_words_dictionary)  # create a deepcopy since we're going to pop()
     for idx, syllables_list in enumerate(word_list):
         for syllable in syllables_list:
             if not syllable.get('has_synalepha'):
                 prosodic_line.append(syllable)
             else:
                 next_syllable = word_list[idx + 1].pop(0)
```
Comment (versae) [inline src/rantanplan/core.py]: I still have some concerns with this implementation. A `for...loop` with a `.pop()` on the control variable.

File: src/rantanplan/core.py [hunk 18]
```diff
@@ -99,69 +99,96 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+    """
+    Gets a list of dictionaries for each word on each line of the poem
+    and joins the syllables to create phonologic syllables according to its
+    'has_synaloepha' value
+    :param sinaereris_words_dictionary: List of dictionaries for each word of the poem
+    :return: A list of conjoined syllables
+    """
     prosodic_line = []
-    word_list = copy.deepcopy(words)
+    word_list = copy.deepcopy(sinaereris_words_dictionary)  # create a deepcopy since we're going to pop()
```
Comment (versae) [inline src/rantanplan/core.py]: Why the change? Why not `words_dict`? Just wondering.

File: src/rantanplan/core.py [hunk 19]
```diff
@@ -99,69 +99,96 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+    or "index": [1,2,4,6] format
+    :param line: a dictionary with the information of the poem
+    :param rhythm_format: The output format for the rhythm
+    :return: Dictionary with with rhythm and phonologic groups
+    """
+    rhythm_dict = {}
+    rhythmical_stress = ""
+    token_list = line['tokens']
+    sinaeresis_tokens = {'sinaeresis_tokens': []}
+    phonologic_groups = {'phonologic_groups': []}
```
Comment (versae) [inline src/rantanplan/core.py]: `phonological_groups = {'phonological_groups': []}` maybe?

File: src/rantanplan/core.py [hunk 20]
```diff
@@ -99,69 +99,96 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+        third_from_last_syllable_stress = last_word[-3].get('is_stressed')
+        if third_from_last_syllable_stress:
+            rhythmical_stress = rhythmical_stress[:-1]
+    elif len(last_word) > 3:  # Stress on preantepenultimate syllable
+        fourth_from_last_syllable_stress = last_word[-4].get('is_stressed')
+        if fourth_from_last_syllable_stress:
+            rhythmical_stress = rhythmical_stress[:-1]
+    if rhythm_format == 'indexed':
+        rhythm_dict = {
+            'rhythm': {'rhythmical_stress': [match.start() for match in re.finditer(r'\+', rhythmical_stress)]},
```
Comment (versae) [inline src/rantanplan/core.py]: Line way too long

File: src/rantanplan/core.py [hunk 21]
```diff
@@ -99,69 +99,96 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+            'type': rhythm_format}
+    elif rhythm_format == 'pattern':
+        rhythm_dict = {'rhythm': {'rhythmical_stress': rhythmical_stress, 'type': rhythm_format}}
```
Comment (versae) [inline src/rantanplan/core.py]: It could use some indentation.

File: src/rantanplan/core.py [hunk 22]
```diff
@@ -266,7 +298,7 @@ def spacy_tag_to_dict(tag):
 
 def get_word_stress(word, pos, tag):
     """
-    Gets a list of syllables from a word and creates a list with syllabified word and stressed syllable index
+    Gets a list of syllables from a ord and creates a list with syllabified word and stressed syllable index
```
Comment (versae) [inline src/rantanplan/core.py]: What's an `ord`?

File: src/rantanplan/core.py [hunk 23]
```diff
@@ -376,18 +409,21 @@ def get_scansion(text, rhyme_analysis=False):
...
-                line["ending_stress"] = rhyme["endings_stress"][index]
-                if line["ending_stress"] == 0:
-                    line["rhyme_type"] = ""
-                    line["rhyme_relaxation"] = None
-                else:
-                    line["rhyme_type"] = rhyme["rhyme_type"]
-                    line["rhyme_relaxation"] = rhyme["rhyme_relaxation"]
+        # Test
+        if analyze_rhyme(lines) is not None:
+            for rhyme in [analyze_rhyme(lines)]:
```
Comment (versae) [inline src/rantanplan/core.py]: No need for that comment I think. Plus, invoking twice `analyze_rhyme(lines)` is a bit wasteful.

File: src/rantanplan/core.py [hunk 24]
```diff
@@ -99,69 +99,96 @@ def have_prosodic_liaison(first_syllable, second_syllable):
...
+        if fourth_from_last_syllable_stress:
+            rhythmical_stress = rhythmical_stress[:-1]
+    if rhythm_format == 'indexed':
+        rhythm_dict = {
+            'rhythm': {'rhythmical_stress': [match.start() for match in re.finditer(r'\+', rhythmical_stress)]},
+            'type': rhythm_format}
+    elif rhythm_format == 'pattern':
+        rhythm_dict = {'rhythm': {'rhythmical_stress': rhythmical_stress, 'type': rhythm_format}}
+    rhythm_dict.update({'rhythmical_length': len(rhythmical_stress)})  # Count the number of phonologic tokens
+    return {**phonologic_groups, **rhythm_dict}
```
Comment (versae) [inline src/rantanplan/core.py]: Neato!

File: src/rantanplan/core.py (hunk 15 above)
Comment (alvp) [reply src/rantanplan/core.py]: We decided to change the name, can't remember why. We also use "rhythmical_pattern"

File: src/rantanplan/core.py (hunk 18 above)
Comment (alvp) [reply src/rantanplan/core.py]: To make clear that we are working with sinaeresis, clarity I guess?

File: src/rantanplan/core.py (hunk 19 above)
Comment (alvp) [reply src/rantanplan/core.py]: Yes, changed

File: src/rantanplan/core.py (hunk 20 above)
Comment (alvp) [reply src/rantanplan/core.py]: Changed a bit, let me know if it's ok

File: src/rantanplan/core.py (hunk 21 above)
Comment (alvp) [reply src/rantanplan/core.py]: Indeed

File: src/rantanplan/core.py (hunk 22 above)
Comment (alvp) [reply src/rantanplan/core.py]: he mallest lement hat an e ttered n solation ith bjective r ractical eaning

Comment (alvp) [reply src/rantanplan/core.py]: fixed

File: src/rantanplan/core.py (hunk 23 above)
Comment (alvp) [reply src/rantanplan/core.py]: Maybe save it to a variable then?

File: src/rantanplan/core.py (hunk 24 above)
Comment (alvp) [reply src/rantanplan/core.py]: <3

File: src/rantanplan/core.py (hunk 23 above)
Comment (alvp) [reply src/rantanplan/core.py]: fixed doing just that, let me know if it's ok

File: src/rantanplan/core.py [hunk 25]
```diff
@@ -72,8 +72,11 @@
 # words starting with OBREP-
 OBREP_RE = re.compile("^(ob)(rep.*)", re.I | re.U)
 
-# words starting with prefixes SIN-/DES- followed by consonant "destituir", "sinhueso"
-PREFIXES_WITH_CONSONANT_RE = (re.compile("^(des|sin)([bcdfgjklmhnñpqrstvxyz].*)", re.I | re.U))
+# words starting with prefixes SIN-/DES- followed by consonant "destituir"
+PREFIX_DES_WITH_CONSONANT_RE = (re.compile("^(des|sin)([bcdfgjklmhnñpqrstvxyz].*)", re.I | re.U))
+
+# words starting with prefixes SIN-/DES- followed by consonant "sinhueso"
+PREFIX_SIN_WITH_CONSONANT_RE = (re.compile("^(des|sin)([bcdfgjklmhnñpqrstvxyz].*)", re.I | re.U))
```
Comment (versae) [inline src/rantanplan/core.py]: If the regular expression is exactly the same, why split it in two variables of identical content?

File: src/rantanplan/core.py [hunk 26]
```diff
@@ -412,9 +427,11 @@ def get_scansion(text, rhyme_analysis=False, rhythm_format="pattern"):
     for line in lines:
         line.update(get_rhythmical_pattern(line, rhythm_format))
     if rhyme_analysis:
-        # Test
-        if analyze_rhyme(lines) is not None:
-            for rhyme in [analyze_rhyme(lines)]:
+        analyzed_lines = analyze_rhyme(lines)
+        print(analyzed_lines)
```
Comment (versae) [inline src/rantanplan/core.py]: Well...

File: src/rantanplan/core.py [hunk 27]
```diff
@@ -412,9 +427,11 @@ def get_scansion(text, rhyme_analysis=False, rhythm_format="pattern"):
...
+        if analyzed_lines is not None:
+            print("caca")
```
Comment (versae) [inline src/rantanplan/core.py]: Seriously?

File: src/rantanplan/core.py (hunk 26 above)
Comment (alvp) [reply src/rantanplan/core.py]: woopsie

File: src/rantanplan/core.py (hunk 27 above)
Comment (alvp) [reply src/rantanplan/core.py]: doopsie

File: src/rantanplan/core.py (hunk 25 above)
Comment (alvp) [reply src/rantanplan/core.py]: Yes, they'll have to be different for future work, changing it now

File: src/rantanplan/core.py (hunk 11 above)
Comment (versae) [reply src/rantanplan/core.py]: Can this be addressed now?

File: src/rantanplan/core.py (hunk 15 above)
Comment (versae) [reply src/rantanplan/core.py]: Some info on why would be handy.

File: src/rantanplan/core.py (hunk 18 above)
Comment (versae) [reply src/rantanplan/core.py]: A bit verbose, IMHO

File: src/rantanplan/core.py (hunk 22 above)
Comment (versae) [reply src/rantanplan/core.py]: That previous message is pretty hard to read.

File: src/rantanplan/core.py (hunk 11 above)
Comment (alvp) [reply src/rantanplan/core.py]: It is, it's now:
```
    for line in lines:
        line.update(get_rhythmical_pattern(line, rhythm_format))
```

File: src/rantanplan/core.py (hunk 15 above)
Comment (alvp) [reply src/rantanplan/core.py]: We changed the name of every "metric_whatever" to "rhytmmical_whatever" so to make it consistent. I don't understand what do you mean about info, info where?

File: src/rantanplan/core.py (hunk 17 above)
Comment (alvp) [reply src/rantanplan/core.py]: It works for now. Maybe we can handle it later.
For what I recall it was the easiest way to solve an extra syllable problem. It also has to do with changing the list elements of the assigned variable that affects the main variable and I didn't want that.

File: src/rantanplan/core.py (hunk 18 above)
Comment (alvp) [reply src/rantanplan/core.py]: changed it to sinaeresis_words

File: src/rantanplan/alternative_syllabification.py [hunk 28]
```diff
@@ -1,3 +1,77 @@
...
+
+Exaxmple: "maître": "maî-tre"
+
+"""
+
+SYLLABIFICATOR_FOREIGN_WORDS_DICT = {
+    "aarónica": "aa-ró-ni-ca",
+    "aarónico": "aa-ró-ni-co",
+    "aaronita": "aa-ro-ni-ta",
+    "ampère": "am-pè-re",
```
Comment (versae) [inline src/rantanplan/alternative_syllabification.py]: Is it not `am-père` in French pronunciation? I guess that list need to be thoroughly review at some point.

File: src/rantanplan/core.py (hunk 17 above)
Comment (versae) [reply src/rantanplan/core.py]: That all can be fixed with indices, which avoids the duplicated data.

File: src/rantanplan/core.py [hunk 29]
```diff
@@ -310,17 +424,21 @@ def get_scansion(text, rhyme_analysis=False):
...
-                line["ending_stress"] = rhyme["endings_stress"][index]
-                if line["ending_stress"] == 0:
-                    line["rhyme_type"] = ""
-                    line["rhyme_relaxation"] = None
-                else:
-                    line["rhyme_type"] = rhyme["rhyme_type"]
-                    line["rhyme_relaxation"] = rhyme["rhyme_relaxation"]
+        analyzed_lines = analyze_rhyme(lines)
+        if analyzed_lines is not None:
+            for rhyme in [analyzed_lines]:
```
Comment (versae) [inline src/rantanplan/core.py]: This is so confusing. Once the PR is done, we should carefully review the code to make it a bit clearer.

File: src/rantanplan/alternative_syllabification.py (hunk 28 above)
Comment (alvp) [reply src/rantanplan/alternative_syllabification.py]: I've done my homework :)
https://dirae.es/palabras/amp%C3%A8re

File: src/rantanplan/core.py (hunk 29 above)
Comment (alvp) [reply src/rantanplan/core.py]: Yes I agree

Comment (versae) [reply src/rantanplan/core.py]: Adding an issue (#55)
